// @routerprotocol/router-js-sdk/dist/router-js-sdk.esm.js
import axios from 'axios';
import { ethers } from 'ethers';
import { Interface } from '@ethersproject/abi';
import { Provider, Contract } from 'ethers-multicall';

var networks = [
	{
		name: "Polygon",
		type: "ethereum",
		networkId: "137",
		id: "1",
		endpoint: "https://polygon-rpc.com",
		socket: "wss://matic-mainnet-archive-ws.bwarelabs.com",
		from: "0x26bC32cD1B7d5Ab87C403C98B4B51AC4EBd7CdA3",
		stakingRewardsFactory: "0xB06224B68309E2F0951E0661bb82A1f42abca7AB",
		native: "0x0000000000000000000000000000000000001010",
		wnative: "0x4c28f48448720e9000907BC2611F73022fdcE1fA",
		rewardsToken: "0xCf4e10cb37907A7b0aDA82aCa5F2E69842dfd6C2",
		feeHandler: "0xA63f3147C6bAE70A53a30e30421b1a992e60Aac2",
		http: "true",
		bridge: "0xf18aCC02628009231d7BAAF9a7a24C0860Dda6cb",
		erc20Handler: "0x00b8F3E184e5DaAadB8fdA995bbabBCC1b35Ae93",
		oneSplit: "0xfEd3c880FF02B195abee916328c5a3953976befD",
		reserveHandler: "0x6e14f48576265272B6CAA3A7cC500a26050Be64E",
		genericHandler: "0x8BFf6bEba0dB9f8104C9b522Cb7b6cAB2B1CCd10",
		gasLimit: "1000000",
		maxGasPrice: "10000000000"
	},
	{
		name: "BSC",
		type: "ethereum",
		networkId: "56",
		id: "2",
		endpoint: "https://bsc-dataseed.binance.org/",
		socket: "wss://bsc-ws-node.nariox.org:443",
		from: "0x26bC32cD1B7d5Ab87C403C98B4B51AC4EBd7CdA3",
		stakingRewardsFactory: "0xB06224B68309E2F0951E0661bb82A1f42abca7AB",
		native: "0x0000000000000000000000000000000000000000",
		wnative: "0xbb4CdB9CBd36B01bD1cBaEBF2De08d9173bc095c",
		rewardsToken: "0xCf4e10cb37907A7b0aDA82aCa5F2E69842dfd6C2",
		feeHandler: "0xA63f3147C6bAE70A53a30e30421b1a992e60Aac2",
		http: "true",
		bridge: "0xf18aCC02628009231d7BAAF9a7a24C0860Dda6cb",
		erc20Handler: "0x00b8F3E184e5DaAadB8fdA995bbabBCC1b35Ae93",
		oneSplit: "0x45d880647Ec9BEF6Bff58ee6bB985C67d7234b0C",
		reserveHandler: "0x6e14f48576265272B6CAA3A7cC500a26050Be64E",
		genericHandler: "0x8BFf6bEba0dB9f8104C9b522Cb7b6cAB2B1CCd10",
		gasLimit: "1000000",
		maxGasPrice: "10000000000"
	},
	{
		name: "Avalanche",
		type: "ethereum",
		networkId: "43114",
		id: "3",
		endpoint: "https://api.avax.network/ext/bc/C/rpc",
		socket: "wss://speedy-nodes-nyc.moralis.io/36a3a9840a5f2cc2ea2bbb42/avalanche/mainnet/ws",
		from: "0x26bC32cD1B7d5Ab87C403C98B4B51AC4EBd7CdA3",
		stakingRewardsFactory: "0xB06224B68309E2F0951E0661bb82A1f42abca7AB",
		native: "0x0100000000000000000000000000000000000001",
		wnative: "0xB31f66AA3C1e785363F0875A1B74E27b85FD66c7",
		rewardsToken: "0xCf4e10cb37907A7b0aDA82aCa5F2E69842dfd6C2",
		feeHandler: "0xA63f3147C6bAE70A53a30e30421b1a992e60Aac2",
		http: "true",
		bridge: "0xf18aCC02628009231d7BAAF9a7a24C0860Dda6cb",
		erc20Handler: "0x00b8F3E184e5DaAadB8fdA995bbabBCC1b35Ae93",
		oneSplit: "0x67b91E05A4Aa940B9892C0dDAAEe0F7483F6B102",
		reserveHandler: "0x6e14f48576265272B6CAA3A7cC500a26050Be64E",
		genericHandler: "0x8BFf6bEba0dB9f8104C9b522Cb7b6cAB2B1CCd10",
		gasLimit: "1000000",
		maxGasPrice: "10000000000"
	},
	{
		name: "Fantom",
		type: "ethereum",
		networkId: "250",
		id: "4",
		endpoint: "https://rpc.ftm.tools/",
		socket: "wss://wsapi.fantom.network/",
		from: "0x33bCDe3dA095C1D2681F3167E5696bf5b7d4F321",
		stakingRewardsFactory: "0xB06224B68309E2F0951E0661bb82A1f42abca7AB",
		native: "0x0100000000000000000000000000000000000001",
		wnative: "0x21be370D5312f44cB42ce377BC9b8a0cEF1A4C83",
		rewardsToken: "0xCf4e10cb37907A7b0aDA82aCa5F2E69842dfd6C2",
		feeHandler: "0xA63f3147C6bAE70A53a30e30421b1a992e60Aac2",
		http: "true",
		bridge: "0xf18aCC02628009231d7BAAF9a7a24C0860Dda6cb",
		erc20Handler: "0x00b8F3E184e5DaAadB8fdA995bbabBCC1b35Ae93",
		oneSplit: "0x621F0549102262148f6a7D289D8330adf7CbC09F",
		reserveHandler: "0x6e14f48576265272B6CAA3A7cC500a26050Be64E",
		genericHandler: "0xA86275db31D136b65E94e475EE16D17e32B24773",
		gasLimit: "1000000",
		maxGasPrice: "10000000000"
	},
	{
		name: "Arbitrum",
		type: "ethereum",
		networkId: "42161",
		id: "5",
		endpoint: "https://arb-mainnet.g.alchemy.com/v2/IhR5U7voO1FpUS3qn-58FYxlPndKtPnk",
		socket: "wss://arb-mainnet.g.alchemy.com/v2/IhR5U7voO1FpUS3qn-58FYxlPndKtPnk",
		from: "0x33bCDe3dA095C1D2681F3167E5696bf5b7d4F321",
		stakingRewardsFactory: "0xB06224B68309E2F0951E0661bb82A1f42abca7AB",
		native: "0x0000000000000000000000000000000000001010",
		wnative: "0x82aF49447D8a07e3bd95BD0d56f35241523fBab1",
		rewardsToken: "0xCf4e10cb37907A7b0aDA82aCa5F2E69842dfd6C2",
		feeHandler: "0xD5b5a08A9Ea195297b6842110229988B8D282A3f",
		http: "true",
		bridge: "0xf18aCC02628009231d7BAAF9a7a24C0860Dda6cb",
		erc20Handler: "0x00b8F3E184e5DaAadB8fdA995bbabBCC1b35Ae93",
		oneSplit: "0x88b1E0ecaC05b876560eF072d51692F53932b16f",
		reserveHandler: "0x6e14f48576265272B6CAA3A7cC500a26050Be64E",
		genericHandler: "0xA86275db31D136b65E94e475EE16D17e32B24773",
		gasLimit: "1000000",
		maxGasPrice: "10000000000"
	},
	{
		name: "Optimism",
		type: "ethereum",
		networkId: "10",
		id: "6",
		endpoint: "https://mainnet.optimism.io",
		socket: "wss://ws-mainnet.optimism.io",
		from: "0x33bCDe3dA095C1D2681F3167E5696bf5b7d4F321",
		stakingRewardsFactory: "0xB06224B68309E2F0951E0661bb82A1f42abca7AB",
		native: "0x0000000000000000000000000000000000001010",
		wnative: "0x4200000000000000000000000000000000000006",
		rewardsToken: "0xCf4e10cb37907A7b0aDA82aCa5F2E69842dfd6C2",
		feeHandler: "0xA63f3147C6bAE70A53a30e30421b1a992e60Aac2",
		http: "true",
		bridge: "0xf18aCC02628009231d7BAAF9a7a24C0860Dda6cb",
		erc20Handler: "0x00b8F3E184e5DaAadB8fdA995bbabBCC1b35Ae93",
		oneSplit: "0xf1684d1DEB902864F2eB0849f50409599322Ce16",
		reserveHandler: "0x6e14f48576265272B6CAA3A7cC500a26050Be64E",
		genericHandler: "0xA86275db31D136b65E94e475EE16D17e32B24773",
		gasLimit: "1000000",
		maxGasPrice: "10000000000"
	},
	{
		name: "Ethereum",
		type: "ethereum",
		networkId: "1",
		id: "7",
		endpoint: "https://eth-mainnet.alchemyapi.io/v2/ZR4W5xz6Q_Ok6Pvpp9e3yIG01MM39zL6",
		socket: "wss://eth-mainnet.alchemyapi.io/v2/ZR4W5xz6Q_Ok6Pvpp9e3yIG01MM39zL6",
		from: "0x33bCDe3dA095C1D2681F3167E5696bf5b7d4F321",
		stakingRewardsFactory: "0xB06224B68309E2F0951E0661bb82A1f42abca7AB",
		native: "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE",
		wnative: "0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2",
		rewardsToken: "0xCf4e10cb37907A7b0aDA82aCa5F2E69842dfd6C2",
		feeHandler: "0xA63f3147C6bAE70A53a30e30421b1a992e60Aac2",
		http: "true",
		bridge: "0xf18aCC02628009231d7BAAF9a7a24C0860Dda6cb",
		erc20Handler: "0x00b8F3E184e5DaAadB8fdA995bbabBCC1b35Ae93",
		oneSplit: "0x5e9A385a15cDE1b149Cb215d9cF3151096A37D67",
		reserveHandler: "0x6e14f48576265272B6CAA3A7cC500a26050Be64E",
		genericHandler: "0xA86275db31D136b65E94e475EE16D17e32B24773",
		gasLimit: "1000000",
		maxGasPrice: "10000000000"
	},
	{
		name: "Harmony",
		type: "ethereum",
		networkId: "1666600000",
		id: "8",
		endpoint: "https://api.harmony.one",
		socket: "wss://ws.s0.t.hmny.io",
		from: "0x33bCDe3dA095C1D2681F3167E5696bf5b7d4F321",
		stakingRewardsFactory: "0xB06224B68309E2F0951E0661bb82A1f42abca7AB",
		native: "0x03fF0ff224f904be3118461335064bB48Df47938",
		wnative: "0xcF664087a5bB0237a0BAd6742852ec6c8d69A27a",
		rewardsToken: "0xCf4e10cb37907A7b0aDA82aCa5F2E69842dfd6C2",
		feeHandler: "0xA63f3147C6bAE70A53a30e30421b1a992e60Aac2",
		http: "true",
		bridge: "0xf18aCC02628009231d7BAAF9a7a24C0860Dda6cb",
		erc20Handler: "0x00b8F3E184e5DaAadB8fdA995bbabBCC1b35Ae93",
		oneSplit: "0x8413041a7702603d9d991F2C4ADd29e4e8A241F8",
		reserveHandler: "0x6e14f48576265272B6CAA3A7cC500a26050Be64E",
		genericHandler: "0xA86275db31D136b65E94e475EE16D17e32B24773",
		gasLimit: "1000000",
		maxGasPrice: "10000000000"
	},
	{
		name: "Aurora",
		type: "ethereum",
		networkId: "1313161554",
		id: "9",
		endpoint: "https://mainnet.aurora.dev",
		socket: "wss://mainnet.aurora.dev",
		from: "0x33bCDe3dA095C1D2681F3167E5696bf5b7d4F321",
		stakingRewardsFactory: "0xB06224B68309E2F0951E0661bb82A1f42abca7AB",
		native: "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE",
		wnative: "0xC9BdeEd33CD01541e1eeD10f90519d2C06Fe3feB",
		rewardsToken: "0xCf4e10cb37907A7b0aDA82aCa5F2E69842dfd6C2",
		feeHandler: "0xA63f3147C6bAE70A53a30e30421b1a992e60Aac2",
		http: "true",
		bridge: "0xf18aCC02628009231d7BAAF9a7a24C0860Dda6cb",
		erc20Handler: "0x00b8F3E184e5DaAadB8fdA995bbabBCC1b35Ae93",
		oneSplit: "0x13538f1450Ca2E1882Df650F87Eb996fF4Ffec34",
		reserveHandler: "0x6e14f48576265272B6CAA3A7cC500a26050Be64E",
		genericHandler: "0xA86275db31D136b65E94e475EE16D17e32B24773",
		gasLimit: "1000000",
		maxGasPrice: "10000000000"
	},
	{
		name: "Cronos",
		type: "ethereum",
		networkId: "25",
		id: "10",
		endpoint: "https://evm.cronos.org",
		socket: "wss://cro.getblock.io/mainnet/7672cfea-5955-4904-9410-6f2c92188075/",
		from: "0x33bCDe3dA095C1D2681F3167E5696bf5b7d4F321",
		stakingRewardsFactory: "0xB06224B68309E2F0951E0661bb82A1f42abca7AB",
		native: "0x0000000000000000000000000000000000000000",
		wnative: "0x5C7F8A570d578ED84E63fdFA7b1eE72dEae1AE23",
		rewardsToken: "0xCf4e10cb37907A7b0aDA82aCa5F2E69842dfd6C2",
		feeHandler: "0xA63f3147C6bAE70A53a30e30421b1a992e60Aac2",
		http: "true",
		bridge: "0xf465FF19D10aE07227130f36fBd34ddE1b264afb",
		erc20Handler: "0x2741986f69Be3D3c92b23DC2f201517939678ba2",
		oneSplit: "0xf44Ff799eA2bBFeC96f9A50498209AAc3C2b3b8b",
		reserveHandler: "0x6e14f48576265272B6CAA3A7cC500a26050Be64E",
		genericHandler: "0xA86275db31D136b65E94e475EE16D17e32B24773",
		gasLimit: "1000000",
		maxGasPrice: "10000000000"
	}
];

const config_generator = () => {
  let network_config = {};
  networks.forEach(network => {
    network_config = { ...network_config,
      [network.networkId]: network
    };
    return;
  });
  return network_config;
};

const NETWORK_CONFIG = /*#__PURE__*/config_generator();

var ropstenAsset = [
	{
		name: "Tether USD",
		symbol: "USDT",
		decimals: "18",
		address: "0x92282AA9861AbE9B987BAd871d5Cc2e4438949fE",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "0x57aF468e958c548B78FC6fc6b603610cB5460B43",
		resourceId: "0x0000000000000000000000b50330607687d414f95384e618c1ae8e2953187600",
		lpResourceId: "",
		chainId: "3",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "USDC",
		symbol: "USDC",
		decimals: "18",
		address: "0xE76CE7d6D51661159164482b6a9cD40c3D93ab8b",
		lpSymbol: "RUSDC",
		lpAddress: "0x0fF7a619bC68c30402062407AB6317Bce69E5c54",
		stakingRewards: "0x57aF468e958c548B78FC6fc6b603610cB5460B43",
		resourceId: "0x0000000000000000000000c25902c1c5c16b9bf73702a8902e26bb0b8da37800",
		lpResourceId: "",
		chainId: "3",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "DAI",
		symbol: "DAI",
		decimals: "18",
		address: "0xfec5f94B8C8a6CBfee8A653bd2366DFcb6b4bF93",
		lpSymbol: "RDAI",
		lpAddress: "0x2ba5a26c4eb686a0d8e43e15afd50978ebd512d5",
		stakingRewards: "",
		resourceId: "0x00000000000000000000002f7b6aa630172313b587ad166004346024e9cfb300",
		lpResourceId: "",
		chainId: "3",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: true
	},
	{
		name: "ETH",
		symbol: "ETH",
		decimals: "18",
		address: "0xc778417E063141139Fce010982780140Aa0cD5Ab",
		lpSymbol: "RETH",
		lpAddress: "0xfc6b7bb453487e86a1bec3f12aee8f56e2ca1c60",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000d0a1e359811322d97991e03f863a0c30c2cf029c00",
		lpResourceId: "",
		chainId: "3",
		logoURI: "",
		mappedOnBridge: true,
		native: true,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: true
	},
	{
		name: "WETH",
		symbol: "WETH",
		decimals: "18",
		address: "0xc778417E063141139Fce010982780140Aa0cD5Ab",
		lpSymbol: "RETH",
		lpAddress: "0xfc6b7bb453487e86a1bec3f12aee8f56e2ca1c60",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000d0a1e359811322d97991e03f863a0c30c2cf029c00",
		lpResourceId: "",
		chainId: "3",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false,
		enableSameChain: true,
		enableCrossChain: false
	},
	{
		name: "ROUTE",
		symbol: "ROUTE",
		decimals: "18",
		address: "0xb073b1002362adF8C81de822DfF2925454499f37",
		lpSymbol: "RROUTE",
		lpAddress: "0x5ae859716fc0ed02280fc4b0c9be79acee78b44a",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000f0e7efed8e866c0de082518e347a4a8d0944abf700",
		lpResourceId: "",
		chainId: "3",
		logoURI: "",
		mappedOnBridge: false,
		native: false,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "DFYN",
		symbol: "DFYN",
		decimals: "18",
		address: "0x81227171CeEC999E79c1CbB50dfc5b4a682c577E",
		lpSymbol: "RDFYN",
		lpAddress: "0x5B59DFe5aD97aEE99Aee2CAEE0963BA8814ae7Fa",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000ccf8c0b61a390dd2f6ff5f910feaccd064b9ac3f00",
		lpResourceId: "",
		chainId: "3",
		logoURI: "",
		mappedOnBridge: false,
		native: false,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	}
];

var kovanAsset = [
	{
		name: "Tether USD",
		symbol: "USDT",
		decimals: "18",
		address: "0xB50330607687d414F95384E618C1ae8E29531876",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000b50330607687d414f95384e618c1ae8e2953187600",
		lpResourceId: "",
		chainId: "42",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: false,
		stableAsset: true,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "USD Coin",
		symbol: "USDC",
		decimals: "18",
		address: "0xc25902c1c5C16b9bF73702A8902E26bB0b8dA378",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000c25902c1c5c16b9bf73702a8902e26bb0b8da37800",
		lpResourceId: "",
		chainId: "42",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "ROUTE",
		symbol: "ROUTE",
		decimals: "18",
		address: "0xf0E7EFED8e866c0DE082518E347a4A8d0944AbF7",
		lpSymbol: "RROUTE",
		lpAddress: "0xd0c68c5158ede68c14bcac2d9c6cf4e36b444065",
		stakingRewards: "",
		resourceId: "",
		lpResourceId: "",
		chainId: "42",
		logoURI: "",
		mappedOnBridge: false,
		native: false,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "DFYN",
		symbol: "DFYN",
		decimals: "18",
		address: "0xccf8C0B61A390dD2F6fF5F910feACCD064b9ac3f",
		lpSymbol: "RDFYN",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x52dc0c3032b82d547d6ffd6482a816eb73b18c71",
		lpResourceId: "",
		chainId: "42",
		logoURI: "",
		mappedOnBridge: false,
		native: false,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "DAI",
		symbol: "DAI",
		decimals: "18",
		address: "0x2f7B6aA630172313b587AD166004346024E9CFb3",
		lpSymbol: "RDAI",
		lpAddress: "0xdf9fe8630309383389ac3f2bdf1e0a7b9db9b04e",
		stakingRewards: "",
		resourceId: "0x00000000000000000000002f7b6aa630172313b587ad166004346024e9cfb300",
		lpResourceId: "",
		chainId: "42",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: true
	},
	{
		name: "ETH",
		symbol: "ETH",
		decimals: "18",
		address: "0xd0A1E359811322d97991E03f863a0C30C2cF029C",
		lpSymbol: "RETH",
		lpAddress: "0xfc6b7bb453487e86a1bec3f12aee8f56e2ca1c60",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000d0a1e359811322d97991e03f863a0c30c2cf029c00",
		lpResourceId: "",
		chainId: "42",
		logoURI: "",
		mappedOnBridge: true,
		native: true,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: true
	},
	{
		name: "WETH",
		symbol: "WETH",
		decimals: "18",
		address: "0xd0A1E359811322d97991E03f863a0C30C2cF029C",
		lpSymbol: "RETH",
		lpAddress: "0xfc6b7bb453487e86a1bec3f12aee8f56e2ca1c60",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000d0a1e359811322d97991e03f863a0c30c2cf029c00",
		lpResourceId: "",
		chainId: "42",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "WMATIC",
		symbol: "WMATIC",
		decimals: "18",
		address: "0xCAE08aa9d3696F50b9096E41558F8DbC6E06aC0d",
		lpSymbol: "RMATIC",
		lpAddress: "0x58380aE0d371BC21e763FA74da75fF84f1a598A3",
		stakingRewards: "",
		resourceId: "0x00000000000000000000006373c962dcffc21465973150993e19f56d8640a400",
		lpResourceId: "",
		chainId: "42",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: true
	},
	{
		name: "Route LP - ROUTE",
		symbol: "RROUTE",
		decimals: "18",
		address: "0xd0c68c5158ede68c14bcac2d9c6cf4e36b444065",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "",
		lpResourceId: "",
		chainId: "42",
		logoURI: "",
		mappedOnBridge: false,
		native: false,
		isLpToken: true,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "Route LP - DFYN",
		symbol: "RDFYN",
		decimals: "18",
		address: "0x52dc0c3032b82d547d6ffd6482a816eb73b18c71",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "",
		lpResourceId: "",
		chainId: "42",
		logoURI: "",
		mappedOnBridge: false,
		native: false,
		isLpToken: true,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "Route LP - DAI",
		symbol: "RDAI",
		decimals: "18",
		address: "0xdf9fe8630309383389ac3f2bdf1e0a7b9db9b04e",
		lpSymbol: "RDAI",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000df9fe8630309383389ac3f2bdf1e0a7b9db9b04e00",
		lpResourceId: "",
		chainId: "42",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: true,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "Route LP - RETH",
		symbol: "RETH",
		decimals: "18",
		address: "0xfc6b7bb453487e86a1bec3f12aee8f56e2ca1c60",
		lpSymbol: "RETH",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000fc6b7bb453487e86a1bec3f12aee8f56e2ca1c6000",
		lpResourceId: "",
		chainId: "42",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: true,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	}
];

var maticAsset = [
	{
		name: "Tether USD",
		symbol: "USDT",
		decimals: "18",
		address: "0x3151Ca2931822CC510aEAB21d74b7c55C3C5dba3",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000b50330607687d414f95384e618c1ae8e2953187600",
		lpResourceId: "",
		chainId: "80001",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: false,
		stableAsset: true,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "USD Coin",
		symbol: "USDC",
		decimals: "18",
		address: "0xe91E3aEfEAd645eea8a9007CEe7F64C0c913285F",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000c25902c1c5c16b9bf73702a8902e26bb0b8da37800",
		lpResourceId: "",
		chainId: "80001",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "ROUTE",
		symbol: "ROUTE",
		decimals: "18",
		address: "0xB0403c233e67EF5f0e424919A7BA19eA53e35939",
		lpSymbol: "RROUTE",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "",
		lpResourceId: "",
		chainId: "80001",
		logoURI: "",
		mappedOnBridge: false,
		native: false,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "DFYN",
		symbol: "DFYN",
		decimals: "18",
		address: "0xB3dBcAd0Af68d8C64267e0B1a8afE4C71aE58009",
		lpSymbol: "RDFYN",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "",
		lpResourceId: "",
		chainId: "80001",
		logoURI: "",
		mappedOnBridge: false,
		native: false,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "DAI",
		symbol: "DAI",
		decimals: "18",
		address: "0x22DF9eF268693fa4875bd58ddf31Ac5e1c3B0DFb",
		lpSymbol: "RDAI",
		lpAddress: "0x961cf7d7cf8bf42831053f1964f5e1d603500678",
		stakingRewards: "",
		resourceId: "0x00000000000000000000002f7b6aa630172313b587ad166004346024e9cfb300",
		lpResourceId: "",
		chainId: "80001",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: true
	},
	{
		name: "WETH",
		symbol: "WETH",
		decimals: "18",
		address: "0x4a67B0fE27399063C3FbFAc4e27926644c6F707e",
		lpSymbol: "RETH",
		lpAddress: "0x386964385e6e0023b9a6af87d2da889f485cd308",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000d0a1e359811322d97991e03f863a0c30c2cf029c00",
		lpResourceId: "",
		chainId: "80001",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: true
	},
	{
		name: "MATIC",
		symbol: "MATIC",
		decimals: "18",
		address: "0x6373c962DCFfc21465973150993E19F56d8640a4",
		lpSymbol: "RMATIC",
		lpAddress: "0xe4b9a77fd07b4807d91ff41692f4e2f4dbd7aa9c",
		stakingRewards: "",
		resourceId: "0x00000000000000000000006373c962dcffc21465973150993e19f56d8640a400",
		lpResourceId: "",
		chainId: "80001",
		logoURI: "",
		mappedOnBridge: true,
		native: true,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: true
	},
	{
		name: "WMATIC",
		symbol: "WMATIC",
		decimals: "18",
		address: "0x6373c962DCFfc21465973150993E19F56d8640a4",
		lpSymbol: "RMATIC",
		lpAddress: "0xe4b9a77fd07b4807d91ff41692f4e2f4dbd7aa9c",
		stakingRewards: "",
		resourceId: "0x00000000000000000000006373c962dcffc21465973150993e19f56d8640a400",
		lpResourceId: "",
		chainId: "80001",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "Route LP - ROUTE",
		symbol: "RROUTE",
		decimals: "18",
		address: "0x579c39b6362cb946ccf44a04d8b5cdda98749302",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "",
		lpResourceId: "",
		chainId: "80001",
		logoURI: "",
		mappedOnBridge: false,
		native: false,
		isLpToken: true,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "Route LP - DFYN",
		symbol: "RDFYN",
		decimals: "18",
		address: "0xba3e2f49de7d833ffdf5064b6be88573035fc36c",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "",
		lpResourceId: "",
		chainId: "80001",
		logoURI: "",
		mappedOnBridge: false,
		native: false,
		isLpToken: true,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "Route LP - DAI",
		symbol: "RDAI",
		decimals: "18",
		address: "0x961cf7d7cf8bf42831053f1964f5e1d603500678",
		lpSymbol: "RDAI",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000df9fe8630309383389ac3f2bdf1e0a7b9db9b04e00",
		lpResourceId: "",
		chainId: "80001",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: true,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "Route LP - RETH",
		symbol: "RETH",
		decimals: "18",
		address: "0x386964385e6e0023b9a6af87d2da889f485cd308",
		lpSymbol: "RETH",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000fc6b7bb453487e86a1bec3f12aee8f56e2ca1c6000",
		lpResourceId: "",
		chainId: "80001",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: true,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	}
];

var bscAsset = [
	{
		name: "BUSD",
		symbol: "BUSD",
		decimals: "18",
		address: "0xe9e7CEA3DedcA5984780Bafc599bD69ADd087D56",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "",
		lpResourceId: "",
		chainId: "56",
		logoURI: "",
		mappedOnBridge: false,
		native: false,
		isLpToken: false,
		hasLpToken: false,
		stableAsset: true,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "USDT",
		symbol: "USDT",
		decimals: "18",
		address: "0x55d398326f99059fF775485246999027B3197955",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "",
		lpResourceId: "",
		chainId: "56",
		logoURI: "",
		mappedOnBridge: false,
		native: false,
		isLpToken: false,
		hasLpToken: false,
		stableAsset: true,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "USDC",
		symbol: "USDC",
		decimals: "18",
		address: "0x8AC76a51cc950d9822D68b83fE1Ad97B32Cd580d",
		lpSymbol: "RUSDC",
		lpAddress: "0x510d0372f55a1687ce73e9826d33d7f3e43d8492",
		stakingRewards: "",
		resourceId: "0x00000000000000000000002791bca1f2de4661ed88a30c99a7a9449aa8417400",
		lpResourceId: "0x0000000000000000000000510d0372f55a1687ce73e9826d33d7f3e43d849200",
		chainId: "56",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: true,
		mining: false,
		activeMining: false,
		enableLiquidityMining: true
	},
	{
		name: "RUSDC",
		symbol: "RUSDC",
		decimals: "18",
		address: "0x510d0372f55a1687ce73e9826d33d7f3e43d8492",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000510d0372f55a1687ce73e9826d33d7f3e43d849200",
		lpResourceId: "",
		chainId: "56",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: true,
		hasLpToken: false,
		stableAsset: true,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "DAI",
		symbol: "DAI",
		decimals: "18",
		address: "0x1AF3F329e8BE154074D8769D1FFa4eE058B1DBc3",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "",
		lpResourceId: "",
		chainId: "56",
		logoURI: "",
		mappedOnBridge: false,
		native: false,
		isLpToken: false,
		hasLpToken: false,
		stableAsset: true,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "ROUTE",
		symbol: "ROUTE",
		decimals: "18",
		address: "0xfD2700c51812753215754De9EC51Cdd42Bf725B9",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x000000000000000000000016eccfdbb4ee1a85a33f3a9b21175cd7ae753db400",
		lpResourceId: "",
		chainId: "56",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "DFYN",
		symbol: "DFYN",
		decimals: "18",
		address: "0x9c08941465EB16982Fa8385A7BbD74F7972C5a27",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000c168e40227e4ebd8c1cae80f7a55a4f0e6d66c9700",
		lpResourceId: "",
		chainId: "56",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "BNB",
		symbol: "BNB",
		decimals: "18",
		address: "0xbb4CdB9CBd36B01bD1cBaEBF2De08d9173bc095c",
		lpSymbol: "RBNB",
		lpAddress: "0x25cf0a692E7736ddB4c45aC33c84c9EC8B0303E4",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000bb4cdb9cbd36b01bd1cbaebf2de08d9173bc095c00",
		lpResourceId: "0x000000000000000000000025cf0a692e7736ddb4c45ac33c84c9ec8b0303e400",
		chainId: "56",
		logoURI: "",
		mappedOnBridge: true,
		native: true,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: true
	},
	{
		name: "WBNB",
		symbol: "WBNB",
		decimals: "18",
		address: "0xbb4CdB9CBd36B01bD1cBaEBF2De08d9173bc095c",
		lpSymbol: "RBNB",
		lpAddress: "0x25cf0a692E7736ddB4c45aC33c84c9EC8B0303E4",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000bb4cdb9cbd36b01bd1cbaebf2de08d9173bc095c00",
		lpResourceId: "0x000000000000000000000025cf0a692e7736ddb4c45ac33c84c9ec8b0303e400",
		chainId: "56",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "WMATIC",
		symbol: "WMATIC",
		decimals: "18",
		address: "0xcc42724c6683b7e57334c4e856f4c9965ed682bd",
		lpSymbol: "RMATIC",
		lpAddress: "0xC3394bB9B118B3c9B5B8dA297e20C08878eDBed4",
		stakingRewards: "",
		resourceId: "0x00000000000000000000004c28f48448720e9000907bc2611f73022fdce1fa00",
		lpResourceId: "0x0000000000000000000000c3394bb9b118b3c9b5b8da297e20c08878edbed400",
		chainId: "56",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: true
	},
	{
		name: "WAVAX",
		symbol: "WAVAX",
		decimals: "18",
		address: "0x1CE0c2827e2eF14D5C4f29a091d735A204794041",
		lpSymbol: "RAVAX",
		lpAddress: "0x1fe37D480C6B1887C4FBCEeB54DF209f90d7a519",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000b31f66aa3c1e785363f0875a1b74e27b85fd66c700",
		lpResourceId: "0x00000000000000000000001fe37d480c6b1887c4fbceeb54df209f90d7a51900",
		chainId: "56",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "WFTM",
		symbol: "WFTM",
		decimals: "18",
		address: "0xAD29AbB318791D579433D831ed122aFeAf29dcfe",
		lpSymbol: "RFTM",
		lpAddress: "0x1fe37D480C6B1887C4FBCEeB54DF209f90d7a519",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000b31f66aa3c1e785363f0875a1b74e27b85fd66c700",
		lpResourceId: "0x00000000000000000000001fe37d480c6b1887c4fbceeb54df209f90d7a51900",
		chainId: "56",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "WETH",
		symbol: "WETH",
		decimals: "18",
		address: "0x2170Ed0880ac9A755fd29B2688956BD959F933F8",
		lpSymbol: "RETH",
		lpAddress: "0x1fe37D480C6B1887C4FBCEeB54DF209f90d7a519",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000b31f66aa3c1e785363f0875a1b74e27b85fd66c700",
		lpResourceId: "0x00000000000000000000001fe37d480c6b1887c4fbceeb54df209f90d7a51900",
		chainId: "56",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "RMATIC",
		symbol: "RMATIC",
		decimals: "18",
		address: "0xC3394bB9B118B3c9B5B8dA297e20C08878eDBed4",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000c3394bb9b118b3c9b5b8da297e20c08878edbed400",
		lpResourceId: "",
		chainId: "56",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: true,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "RBNB",
		symbol: "RBNB",
		decimals: "18",
		address: "0x25cf0a692E7736ddB4c45aC33c84c9EC8B0303E4",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x000000000000000000000025cf0a692e7736ddb4c45ac33c84c9ec8b0303e400",
		lpResourceId: "",
		chainId: "56",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: true,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "RAVAX",
		symbol: "RAVAX",
		decimals: "18",
		address: "0x1fe37D480C6B1887C4FBCEeB54DF209f90d7a519",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x00000000000000000000001fe37d480c6b1887c4fbceeb54df209f90d7a51900",
		lpResourceId: "",
		chainId: "56",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: true,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "RFTM",
		symbol: "RFTM",
		decimals: "18",
		address: "0x99f7673C449d325Cac340Dd1a45F3B9bc3C67Ce0",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x000000000000000000000099f7673c449d325cac340dd1a45f3b9bc3c67ce000",
		lpResourceId: "",
		chainId: "56",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: true,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "RETH",
		symbol: "RETH",
		decimals: "18",
		address: "0x64FFCaECB9a94569f2E082B260d2794168B96ac7",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x000000000000000000000064ffcaecb9a94569f2e082b260d2794168b96ac700",
		lpResourceId: "",
		chainId: "56",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: true,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	}
];

var avaxFujiAsset = [
	{
		name: "Tether USD",
		symbol: "USDT",
		decimals: "18",
		address: "0xf394Af2D0e10e2C9672Fb155219dFA68F68e04AD",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000b50330607687d414f95384e618c1ae8e2953187600",
		lpResourceId: "",
		chainId: "43113",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: false,
		stableAsset: true,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "USD Coin",
		symbol: "USDC",
		decimals: "18",
		address: "0x685A94e520CfA0AdF6b48dF42Fb1AeF8C024EAf5",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000c25902c1c5c16b9bf73702a8902e26bb0b8da37800",
		lpResourceId: "",
		chainId: "43113",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "AVAX",
		symbol: "AVAX",
		decimals: "18",
		address: "0xd00ae08403B9bbb9124bB305C09058E32C39A48c",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x00000000000000000000006373c962dcffc21465973150993e19f56d8640a400",
		lpResourceId: "",
		chainId: "43113",
		logoURI: "",
		mappedOnBridge: true,
		native: true,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: true
	}
];

var avaxAsset = [
	{
		name: "USDT",
		symbol: "USDT",
		decimals: "6",
		address: "0xc7198437980c041c805A1EDcbA50c1Ce5db95118",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "",
		lpResourceId: "",
		chainId: "43114",
		logoURI: "",
		mappedOnBridge: false,
		native: false,
		isLpToken: false,
		hasLpToken: false,
		stableAsset: true,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "USDC.e",
		symbol: "USDC.e",
		decimals: "6",
		address: "0xA7D7079b0FEaD91F3e65f86E8915Cb59c1a4C664",
		lpSymbol: "RUSDC",
		lpAddress: "0x25cf0a692E7736ddB4c45aC33c84c9EC8B0303E4",
		stakingRewards: "",
		resourceId: "0x00000000000000000000002791bca1f2de4661ed88a30c99a7a9449aa8417400",
		lpResourceId: "0x0000000000000000000000510d0372f55a1687ce73e9826d33d7f3e43d849200",
		chainId: "43114",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: true,
		mining: false,
		activeMining: false,
		enableLiquidityMining: true
	},
	{
		name: "DAI",
		symbol: "DAI",
		decimals: "18",
		address: "0xd586E7F844cEa2F87f50152665BCbc2C279D8d70",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "",
		lpResourceId: "",
		chainId: "43114",
		logoURI: "",
		mappedOnBridge: false,
		native: false,
		isLpToken: false,
		hasLpToken: false,
		stableAsset: true,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "ROUTE",
		symbol: "ROUTE",
		decimals: "18",
		address: "0xf44Ff799eA2bBFeC96f9A50498209AAc3C2b3b8b",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x000000000000000000000016eccfdbb4ee1a85a33f3a9b21175cd7ae753db400",
		lpResourceId: "",
		chainId: "56",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "DFYN",
		symbol: "DFYN",
		decimals: "18",
		address: "0x2b6b20ceaC439042edC6eFD3d41b11af66013e58",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000c168e40227e4ebd8c1cae80f7a55a4f0e6d66c9700",
		lpResourceId: "",
		chainId: "56",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "AVAX",
		symbol: "AVAX",
		decimals: "18",
		address: "0xB31f66AA3C1e785363F0875A1B74E27b85FD66c7",
		lpSymbol: "RAVAX",
		lpAddress: "0x1fe37D480C6B1887C4FBCEeB54DF209f90d7a519",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000b31f66aa3c1e785363f0875a1b74e27b85fd66c700",
		lpResourceId: "0x00000000000000000000001fe37d480c6b1887c4fbceeb54df209f90d7a51900",
		chainId: "43114",
		logoURI: "",
		mappedOnBridge: true,
		native: true,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: true
	},
	{
		name: "WAVAX",
		symbol: "WAVAX",
		decimals: "18",
		address: "0xB31f66AA3C1e785363F0875A1B74E27b85FD66c7",
		lpSymbol: "RAVAX",
		lpAddress: "0x1fe37D480C6B1887C4FBCEeB54DF209f90d7a519",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000b31f66aa3c1e785363f0875a1b74e27b85fd66c700",
		lpResourceId: "0x00000000000000000000001fe37d480c6b1887c4fbceeb54df209f90d7a51900",
		chainId: "43114",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "RAVAX",
		symbol: "RAVAX",
		decimals: "18",
		address: "0x1fe37D480C6B1887C4FBCEeB54DF209f90d7a519",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x00000000000000000000001fe37d480c6b1887c4fbceeb54df209f90d7a51900",
		lpResourceId: "",
		chainId: "43114",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: true,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "WBNB",
		symbol: "WBNB",
		decimals: "18",
		address: "0x264c1383EA520f73dd837F915ef3a732e204a493",
		lpSymbol: "RBNB",
		lpAddress: "0xC3394bB9B118B3c9B5B8dA297e20C08878eDBed4",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000bb4cdb9cbd36b01bd1cbaebf2de08d9173bc095c00",
		lpResourceId: "0x000000000000000000000025cf0a692e7736ddb4c45ac33c84c9ec8b0303e400",
		chainId: "43114",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: true
	},
	{
		name: "WMATIC",
		symbol: "WMATIC",
		decimals: "18",
		address: "0xF23f98fFfa44905991ea74454B1CA8251f6721A0",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x00000000000000000000004c28f48448720e9000907bc2611f73022fdce1fa00",
		lpResourceId: "",
		chainId: "43114",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "WETH",
		symbol: "WETH",
		decimals: "18",
		address: "0x49D5c2BdFfac6CE2BFdB6640F4F80f226bc10bAB",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x00000000000000000000004c28f48448720e9000907bc2611f73022fdce1fa00",
		lpResourceId: "",
		chainId: "43114",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "RUSDC",
		symbol: "RUSDC",
		decimals: "18",
		address: "0x25cf0a692E7736ddB4c45aC33c84c9EC8B0303E4",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000510d0372f55a1687ce73e9826d33d7f3e43d849200",
		lpResourceId: "",
		chainId: "43114",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: true,
		hasLpToken: false,
		stableAsset: true,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "RBNB",
		symbol: "RBNB",
		decimals: "18",
		address: "0xC3394bB9B118B3c9B5B8dA297e20C08878eDBed4",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x000000000000000000000025cf0a692e7736ddb4c45ac33c84c9ec8b0303e400",
		lpResourceId: "",
		chainId: "43114",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: true,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "RETH",
		symbol: "RETH",
		decimals: "18",
		address: "0x99f7673C449d325Cac340Dd1a45F3B9bc3C67Ce0",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x000000000000000000000064ffcaecb9a94569f2e082b260d2794168b96ac700",
		lpResourceId: "",
		chainId: "43114",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: true,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	}
];

var okexAsset = [
	{
		name: "DFYN",
		symbol: "DFYN",
		decimals: "18",
		address: "0x1512Ee157Fd9e74e97dCd8Ffa0C71d23ABad6701",
		lpSymbol: "RDFYN",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000B81b3eCd19820521B36Ee1644b9A17c90d19D24600",
		lpResourceId: "",
		chainId: "65",
		logoURI: "",
		enabled: true,
		mappedOnBridge: false,
		native: false,
		isLpToken: false,
		hasLpToken: false,
		stableAsset: true,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	}
];

var ethereumAsset = [
	{
		name: "USDT",
		symbol: "USDT",
		decimals: "6",
		address: "0xdAC17F958D2ee523a2206206994597C13D831ec7",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "",
		lpResourceId: "",
		chainId: "1",
		logoURI: "",
		mappedOnBridge: false,
		native: false,
		isLpToken: false,
		hasLpToken: false,
		stableAsset: true,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "USDC",
		symbol: "USDC",
		decimals: "6",
		address: "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48",
		lpSymbol: "RUSDC",
		lpAddress: "0x7bc1b2620fbc68c7435c07c812b2cd50b1d98ab9",
		stakingRewards: "",
		resourceId: "0x00000000000000000000002791bca1f2de4661ed88a30c99a7a9449aa8417400",
		lpResourceId: "0x0000000000000000000000510d0372f55a1687ce73e9826d33d7f3e43d849200",
		chainId: "1",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: true,
		mining: false,
		activeMining: false,
		enableLiquidityMining: true
	},
	{
		name: "RUSDC",
		symbol: "RUSDC",
		decimals: "6",
		address: "0x7bc1b2620fbc68c7435c07c812b2cd50b1d98ab9",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000510d0372f55a1687ce73e9826d33d7f3e43d849200",
		lpResourceId: "",
		chainId: "1",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: true,
		hasLpToken: false,
		stableAsset: true,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "DAI",
		symbol: "DAI",
		decimals: "18",
		address: "0x6B175474E89094C44Da98b954EedeAC495271d0F",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "",
		lpResourceId: "",
		chainId: "1",
		logoURI: "",
		mappedOnBridge: false,
		native: false,
		isLpToken: false,
		hasLpToken: false,
		stableAsset: true,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "ROUTE",
		symbol: "ROUTE",
		decimals: "18",
		address: "0x16ECCfDbb4eE1A85A33f3A9B21175Cd7Ae753dB4",
		lpSymbol: "RROUTE",
		lpAddress: "0x64ffcaecb9a94569f2e082b260d2794168b96ac7",
		stakingRewards: "",
		resourceId: "0x000000000000000000000016eccfdbb4ee1a85a33f3a9b21175cd7ae753db400",
		lpResourceId: "0x0000000000000000000000bd9a8c58dc0f99007c0e67f432214cd55907834d00",
		chainId: "1",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: true
	},
	{
		name: "DFYN",
		symbol: "DFYN",
		decimals: "18",
		address: "0x9695e0114e12C0d3A3636fAb5A18e6b737529023",
		lpSymbol: "RDFYN",
		lpAddress: "0xbd9a8c58dc0f99007c0e67f432214cd55907834d",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000c168e40227e4ebd8c1cae80f7a55a4f0e6d66c9700",
		lpResourceId: "0x0000000000000000000000da913f2f57a4a1110bfb6ccbd4e3c88b9ca662f000",
		chainId: "1",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: true
	},
	{
		name: "ETH",
		symbol: "ETH",
		decimals: "18",
		address: "0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2",
		lpSymbol: "RETH",
		lpAddress: "0xc3394bb9b118b3c9b5b8da297e20c08878edbed4",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000c02aaa39b223fe8d0a0e5c4f27ead9083c756cc200",
		lpResourceId: "0x000000000000000000000064ffcaecb9a94569f2e082b260d2794168b96ac700",
		chainId: "1",
		logoURI: "",
		mappedOnBridge: true,
		native: true,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: true
	},
	{
		name: "WETH",
		symbol: "WETH",
		decimals: "18",
		address: "0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2",
		lpSymbol: "RETH",
		lpAddress: "0xc3394bb9b118b3c9b5b8da297e20c08878edbed4",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000c02aaa39b223fe8d0a0e5c4f27ead9083c756cc200",
		lpResourceId: "0x000000000000000000000064ffcaecb9a94569f2e082b260d2794168b96ac700",
		chainId: "1",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "WFTM",
		symbol: "WFTM",
		decimals: "18",
		address: "0x4E15361FD6b4BB609Fa63C81A2be19d873717870",
		lpSymbol: "RFTM",
		lpAddress: "0x510d0372f55a1687ce73e9826d33d7f3e43d8492",
		stakingRewards: "",
		resourceId: "0x000000000000000000000021be370d5312f44cb42ce377bc9b8a0cef1a4c8300",
		lpResourceId: "0x000000000000000000000099f7673c449d325cac340dd1a45f3b9bc3c67ce000",
		chainId: "1",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: true
	},
	{
		name: "WMATIC",
		symbol: "WMATIC",
		decimals: "18",
		address: "0x7D1AfA7B718fb893dB30A3aBc0Cfc608AaCfeBB0",
		lpSymbol: "RMATIC",
		lpAddress: "0x1fe37d480c6b1887c4fbceeb54df209f90d7a519",
		stakingRewards: "",
		resourceId: "0x00000000000000000000004c28f48448720e9000907bc2611f73022fdce1fa00",
		lpResourceId: "0x0000000000000000000000c3394bb9b118b3c9b5b8da297e20c08878edbed400",
		chainId: "1",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: true
	},
	{
		name: "WBNB",
		symbol: "WBNB",
		decimals: "18",
		address: "0xB8c77482e45F1F44dE1745F52C74426C631bDD52",
		lpSymbol: "RBNB",
		lpAddress: "0x25cf0a692e7736ddb4c45ac33c84c9ec8b0303e4",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000bb4cdb9cbd36b01bd1cbaebf2de08d9173bc095c00",
		lpResourceId: "0x000000000000000000000025cf0a692e7736ddb4c45ac33c84c9ec8b0303e400",
		chainId: "1",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: true
	},
	{
		name: "WAVAX",
		symbol: "WAVAX",
		decimals: "18",
		address: "0x85f138bfEE4ef8e540890CFb48F620571d67Eda3",
		lpSymbol: "RAVAX",
		lpAddress: "0x99f7673c449d325cac340dd1a45f3b9bc3c67ce0",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000b31f66aa3c1e785363f0875a1b74e27b85fd66c700",
		lpResourceId: "0x00000000000000000000001fe37d480c6b1887c4fbceeb54df209f90d7a51900",
		chainId: "1",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: true
	},
	{
		name: "RMATIC",
		symbol: "RMATIC",
		decimals: "18",
		address: "0x1fe37d480c6b1887c4fbceeb54df209f90d7a519",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000c3394bb9b118b3c9b5b8da297e20c08878edbed400",
		lpResourceId: "",
		chainId: "1",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: true,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "RBNB",
		symbol: "RBNB",
		decimals: "18",
		address: "0x25cf0a692e7736ddb4c45ac33c84c9ec8b0303e4",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x000000000000000000000025cf0a692e7736ddb4c45ac33c84c9ec8b0303e400",
		lpResourceId: "",
		chainId: "1",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: true,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "RAVAX",
		symbol: "RAVAX",
		decimals: "18",
		address: "0x99f7673c449d325cac340dd1a45f3b9bc3c67ce0",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x00000000000000000000001fe37d480c6b1887c4fbceeb54df209f90d7a51900",
		lpResourceId: "",
		chainId: "1",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: true,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "RFTM",
		symbol: "RFTM",
		decimals: "18",
		address: "0x510d0372f55a1687ce73e9826d33d7f3e43d8492",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x000000000000000000000099f7673c449d325cac340dd1a45f3b9bc3c67ce000",
		lpResourceId: "",
		chainId: "1",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: true,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "RETH",
		symbol: "RETH",
		decimals: "18",
		address: "0xc3394bb9b118b3c9b5b8da297e20c08878edbed4",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x000000000000000000000064ffcaecb9a94569f2e082b260d2794168b96ac700",
		lpResourceId: "",
		chainId: "1",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: true,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "RROUTE",
		symbol: "RROUTE",
		decimals: "18",
		address: "0x64ffcaecb9a94569f2e082b260d2794168b96ac7",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000bd9a8c58dc0f99007c0e67f432214cd55907834d00",
		lpResourceId: "",
		chainId: "1",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: true,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "RDFYN",
		symbol: "RDFYN",
		decimals: "18",
		address: "0xbd9a8c58dc0f99007c0e67f432214cd55907834d",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000da913f2f57a4a1110bfb6ccbd4e3c88b9ca662f000",
		lpResourceId: "",
		chainId: "1",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: true,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	}
];

var polgonAsset = [
	{
		name: "USDT",
		symbol: "USDT",
		decimals: "6",
		address: "0xc2132d05d31c914a87c6611c10748aeb04b58e8f",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "",
		lpResourceId: "",
		chainId: "137",
		logoURI: "",
		mappedOnBridge: false,
		native: false,
		isLpToken: false,
		hasLpToken: false,
		stableAsset: true,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "USDC",
		symbol: "USDC",
		decimals: "6",
		address: "0x2791Bca1f2de4661ED88A30C99A7a9449Aa84174",
		lpSymbol: "RUSDC",
		lpAddress: "0x510d0372F55A1687CE73e9826d33d7f3E43D8492",
		stakingRewards: "",
		resourceId: "0x00000000000000000000002791bca1f2de4661ed88a30c99a7a9449aa8417400",
		lpResourceId: "0x0000000000000000000000510d0372f55a1687ce73e9826d33d7f3e43d849200",
		chainId: "137",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: true,
		mining: false,
		activeMining: false,
		enableLiquidityMining: true
	},
	{
		name: "RUSDC",
		symbol: "RUSDC",
		decimals: "6",
		address: "0x510d0372F55A1687CE73e9826d33d7f3E43D8492",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000510d0372f55a1687ce73e9826d33d7f3e43d849200",
		lpResourceId: "",
		chainId: "137",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: true,
		hasLpToken: false,
		stableAsset: true,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "DAI",
		symbol: "DAI",
		decimals: "18",
		address: "0x8f3Cf7ad23Cd3CaDbD9735AFf958023239c6A063",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "",
		lpResourceId: "",
		chainId: "137",
		logoURI: "",
		mappedOnBridge: false,
		native: false,
		isLpToken: false,
		hasLpToken: false,
		stableAsset: true,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "ROUTE",
		symbol: "ROUTE",
		decimals: "18",
		address: "0x16ECCfDbb4eE1A85A33f3A9B21175Cd7Ae753dB4",
		lpSymbol: "RROUTE",
		lpAddress: "0x64ffcaecb9a94569f2e082b260d2794168b96ac7",
		stakingRewards: "",
		resourceId: "0x000000000000000000000016eccfdbb4ee1a85a33f3a9b21175cd7ae753db400",
		lpResourceId: "0x0000000000000000000000bd9a8c58dc0f99007c0e67f432214cd55907834d00",
		chainId: "137",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: true
	},
	{
		name: "DFYN",
		symbol: "DFYN",
		decimals: "18",
		address: "0xC168E40227E4ebD8C1caE80F7a55a4F0e6D66C97",
		lpSymbol: "RDFYN",
		lpAddress: "0xda913f2f57a4a1110bfb6ccbd4e3c88b9ca662f0",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000c168e40227e4ebd8c1cae80f7a55a4f0e6d66c9700",
		lpResourceId: "0x0000000000000000000000da913f2f57a4a1110bfb6ccbd4e3c88b9ca662f000",
		chainId: "137",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: true
	},
	{
		name: "MATIC",
		symbol: "MATIC",
		decimals: "18",
		address: "0x4c28f48448720e9000907BC2611F73022fdcE1fA",
		lpSymbol: "RMATIC",
		lpAddress: "0xc3394bb9b118b3c9b5b8da297e20c08878edbed4",
		stakingRewards: "",
		resourceId: "0x00000000000000000000004c28f48448720e9000907bc2611f73022fdce1fa00",
		lpResourceId: "0x0000000000000000000000c3394bb9b118b3c9b5b8da297e20c08878edbed400",
		chainId: "137",
		logoURI: "",
		mappedOnBridge: true,
		native: true,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: true
	},
	{
		name: "WMATIC",
		symbol: "WMATIC",
		decimals: "18",
		address: "0x4c28f48448720e9000907BC2611F73022fdcE1fA",
		lpSymbol: "RMATIC",
		lpAddress: "0xc3394bb9b118b3c9b5b8da297e20c08878edbed4",
		stakingRewards: "",
		resourceId: "0x00000000000000000000004c28f48448720e9000907bc2611f73022fdce1fa00",
		lpResourceId: "0x0000000000000000000000c3394bb9b118b3c9b5b8da297e20c08878edbed400",
		chainId: "137",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "WBNB",
		symbol: "WBNB",
		decimals: "18",
		address: "0x5c4b7CCBF908E64F32e12c6650ec0C96d717f03F",
		lpSymbol: "RBNB",
		lpAddress: "0x25cf0a692E7736ddB4c45aC33c84c9EC8B0303E4",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000bb4cdb9cbd36b01bd1cbaebf2de08d9173bc095c00",
		lpResourceId: "0x000000000000000000000025cf0a692e7736ddb4c45ac33c84c9ec8b0303e400",
		chainId: "137",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: true
	},
	{
		name: "WAVAX",
		symbol: "WAVAX",
		decimals: "18",
		address: "0x2C89bbc92BD86F8075d1DEcc58C7F4E0107f286b",
		lpSymbol: "RAVAX",
		lpAddress: "0x1fe37D480C6B1887C4FBCEeB54DF209f90d7a519",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000b31f66aa3c1e785363f0875a1b74e27b85fd66c700",
		lpResourceId: "0x00000000000000000000001fe37d480c6b1887c4fbceeb54df209f90d7a51900",
		chainId: "137",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: true
	},
	{
		name: "WFTM",
		symbol: "WFTM",
		decimals: "18",
		address: "0xC9c1c1c20B3658F8787CC2FD702267791f224Ce1",
		lpSymbol: "RFTM",
		lpAddress: "0x1fe37D480C6B1887C4FBCEeB54DF209f90d7a519",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000b31f66aa3c1e785363f0875a1b74e27b85fd66c700",
		lpResourceId: "0x00000000000000000000001fe37d480c6b1887c4fbceeb54df209f90d7a51900",
		chainId: "137",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: true
	},
	{
		name: "WETH",
		symbol: "WETH",
		decimals: "18",
		address: "0x7ceB23fD6bC0adD59E62ac25578270cFf1b9f619",
		lpSymbol: "RETH",
		lpAddress: "0x1fe37D480C6B1887C4FBCEeB54DF209f90d7a519",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000b31f66aa3c1e785363f0875a1b74e27b85fd66c700",
		lpResourceId: "0x00000000000000000000001fe37d480c6b1887c4fbceeb54df209f90d7a51900",
		chainId: "137",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: true
	},
	{
		name: "RMATIC",
		symbol: "RMATIC",
		decimals: "18",
		address: "0xC3394bB9B118B3c9B5B8dA297e20C08878eDBed4",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000c3394bb9b118b3c9b5b8da297e20c08878edbed400",
		lpResourceId: "",
		chainId: "137",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: true,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "RBNB",
		symbol: "RBNB",
		decimals: "18",
		address: "0x25cf0a692E7736ddB4c45aC33c84c9EC8B0303E4",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x000000000000000000000025cf0a692e7736ddb4c45ac33c84c9ec8b0303e400",
		lpResourceId: "",
		chainId: "137",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: true,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "RAVAX",
		symbol: "RAVAX",
		decimals: "18",
		address: "0x1fe37D480C6B1887C4FBCEeB54DF209f90d7a519",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x00000000000000000000001fe37d480c6b1887c4fbceeb54df209f90d7a51900",
		lpResourceId: "",
		chainId: "137",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: true,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "RETH",
		symbol: "RETH",
		decimals: "18",
		address: "0x1fe37D480C6B1887C4FBCEeB54DF209f90d7a519",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x00000000000000000000001fe37d480c6b1887c4fbceeb54df209f90d7a51900",
		lpResourceId: "",
		chainId: "137",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: true,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "RFTM",
		symbol: "RFTM",
		decimals: "18",
		address: "0x1fe37D480C6B1887C4FBCEeB54DF209f90d7a519",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x00000000000000000000001fe37d480c6b1887c4fbceeb54df209f90d7a51900",
		lpResourceId: "",
		chainId: "137",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: true,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "RROUTE",
		symbol: "RROUTE",
		decimals: "18",
		address: "0xbd9a8c58dc0f99007c0e67f432214cd55907834d",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000bd9a8c58dc0f99007c0e67f432214cd55907834d00",
		lpResourceId: "",
		chainId: "137",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: true,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "RDFYN",
		symbol: "RDFYN",
		decimals: "18",
		address: "0xda913f2f57a4a1110bfb6ccbd4e3c88b9ca662f0",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000da913f2f57a4a1110bfb6ccbd4e3c88b9ca662f000",
		lpResourceId: "",
		chainId: "137",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: true,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	}
];

var ftmAsset = [
	{
		name: "USDT",
		symbol: "USDT",
		decimals: "6",
		address: "0x049d68029688eabf473097a2fc38ef61633a3c7a",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "",
		lpResourceId: "",
		chainId: "250",
		logoURI: "",
		mappedOnBridge: false,
		native: false,
		isLpToken: false,
		hasLpToken: false,
		stableAsset: true,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "USDC",
		symbol: "USDC",
		decimals: "6",
		address: "0x04068DA6C83AFCFA0e13ba15A6696662335D5B75",
		lpSymbol: "RUSDC",
		lpAddress: "0x7bc1b2620fbc68c7435c07c812b2cd50b1d98ab9",
		stakingRewards: "",
		resourceId: "0x00000000000000000000002791bca1f2de4661ed88a30c99a7a9449aa8417400",
		lpResourceId: "0x0000000000000000000000510d0372f55a1687ce73e9826d33d7f3e43d849200",
		chainId: "250",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: true,
		mining: false,
		activeMining: false,
		enableLiquidityMining: true
	},
	{
		name: "RUSDC",
		symbol: "RUSDC",
		decimals: "6",
		address: "0x7bc1b2620fbc68c7435c07c812b2cd50b1d98ab9",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000510d0372f55a1687ce73e9826d33d7f3e43d849200",
		lpResourceId: "",
		chainId: "250",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: true,
		hasLpToken: false,
		stableAsset: true,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "DAI",
		symbol: "DAI",
		decimals: "18",
		address: "0x8D11eC38a3EB5E956B052f67Da8Bdc9bef8Abf3E",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "",
		lpResourceId: "",
		chainId: "250",
		logoURI: "",
		mappedOnBridge: false,
		native: false,
		isLpToken: false,
		hasLpToken: false,
		stableAsset: true,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "ROUTE",
		symbol: "ROUTE",
		decimals: "18",
		address: "0x11BbF12363dC8375b78D2719395d505f52a02F68",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x000000000000000000000016eccfdbb4ee1a85a33f3a9b21175cd7ae753db400",
		lpResourceId: "",
		chainId: "250",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "DFYN",
		symbol: "DFYN",
		decimals: "18",
		address: "0x13538f1450Ca2E1882Df650F87Eb996fF4Ffec34",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000c168e40227e4ebd8c1cae80f7a55a4f0e6d66c9700",
		lpResourceId: "",
		chainId: "250",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "FTM",
		symbol: "FTM",
		decimals: "18",
		address: "0x21be370D5312f44cB42ce377BC9b8a0cEF1A4C83",
		lpSymbol: "RFTM",
		lpAddress: "0x510d0372f55a1687ce73e9826d33d7f3e43d8492",
		stakingRewards: "",
		resourceId: "0x000000000000000000000021be370d5312f44cb42ce377bc9b8a0cef1a4c8300",
		lpResourceId: "0x000000000000000000000099f7673c449d325cac340dd1a45f3b9bc3c67ce000",
		chainId: "250",
		logoURI: "",
		mappedOnBridge: true,
		native: true,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: true
	},
	{
		name: "WFTM",
		symbol: "WFTM",
		decimals: "18",
		address: "0x21be370D5312f44cB42ce377BC9b8a0cEF1A4C83",
		lpSymbol: "RFTM",
		lpAddress: "0x510d0372f55a1687ce73e9826d33d7f3e43d8492",
		stakingRewards: "",
		resourceId: "0x000000000000000000000021be370d5312f44cb42ce377bc9b8a0cef1a4c8300",
		lpResourceId: "0x000000000000000000000099f7673c449d325cac340dd1a45f3b9bc3c67ce000",
		chainId: "250",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "WMATIC",
		symbol: "WMATIC",
		decimals: "18",
		address: "0x40df1ae6074c35047bff66675488aa2f9f6384f3",
		lpSymbol: "RMATIC",
		lpAddress: "0x25cf0a692e7736ddb4c45ac33c84c9ec8b0303e4",
		stakingRewards: "",
		resourceId: "0x00000000000000000000004c28f48448720e9000907bc2611f73022fdce1fa00",
		lpResourceId: "0x0000000000000000000000c3394bb9b118b3c9b5b8da297e20c08878edbed400",
		chainId: "250",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "WBNB",
		symbol: "WBNB",
		decimals: "18",
		address: "0x27f26F00e1605903645BbaBC0a73E35027Dccd45",
		lpSymbol: "RBNB",
		lpAddress: "0xc3394bb9b118b3c9b5b8da297e20c08878edbed4",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000bb4cdb9cbd36b01bd1cbaebf2de08d9173bc095c00",
		lpResourceId: "0x000000000000000000000025cf0a692e7736ddb4c45ac33c84c9ec8b0303e400",
		chainId: "250",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: true
	},
	{
		name: "WAVAX",
		symbol: "WAVAX",
		decimals: "18",
		address: "0x511D35c52a3C244E7b8bd92c0C297755FbD89212",
		lpSymbol: "RAVAX",
		lpAddress: "0x1fe37d480c6b1887c4fbceeb54df209f90d7a519",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000b31f66aa3c1e785363f0875a1b74e27b85fd66c700",
		lpResourceId: "0x00000000000000000000001fe37d480c6b1887c4fbceeb54df209f90d7a51900",
		chainId: "250",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: true
	},
	{
		name: "WETH",
		symbol: "WETH",
		decimals: "18",
		address: "0x74b23882a30290451A17c44f4F05243b6b58C76d",
		lpSymbol: "RAVAX",
		lpAddress: "0x99f7673c449d325cac340dd1a45f3b9bc3c67ce0",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000c02aaa39b223fe8d0a0e5c4f27ead9083c756cc200",
		lpResourceId: "0x000000000000000000000064ffcaecb9a94569f2e082b260d2794168b96ac700",
		chainId: "250",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: true
	},
	{
		name: "RMATIC",
		symbol: "RMATIC",
		decimals: "18",
		address: "0x40DF1Ae6074C35047BFF66675488Aa2f9f6384F3",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000c3394bb9b118b3c9b5b8da297e20c08878edbed400",
		lpResourceId: "",
		chainId: "250",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: true,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "RBNB",
		symbol: "RBNB",
		decimals: "18",
		address: "0xc3394bb9b118b3c9b5b8da297e20c08878edbed4",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x000000000000000000000025cf0a692e7736ddb4c45ac33c84c9ec8b0303e400",
		lpResourceId: "",
		chainId: "250",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: true,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "RAVAX",
		symbol: "RAVAX",
		decimals: "18",
		address: "0x1fe37d480c6b1887c4fbceeb54df209f90d7a519",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x00000000000000000000001fe37d480c6b1887c4fbceeb54df209f90d7a51900",
		lpResourceId: "",
		chainId: "250",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: true,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "RFTM",
		symbol: "RFTM",
		decimals: "18",
		address: "0x510d0372f55a1687ce73e9826d33d7f3e43d8492",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x000000000000000000000099f7673c449d325cac340dd1a45f3b9bc3c67ce000",
		lpResourceId: "",
		chainId: "250",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: true,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "RETH",
		symbol: "RETH",
		decimals: "18",
		address: "0x99f7673c449d325cac340dd1a45f3b9bc3c67ce0",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x000000000000000000000064ffcaecb9a94569f2e082b260d2794168b96ac700",
		lpResourceId: "",
		chainId: "250",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: true,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	}
];

var arbitrumAsset = [
	{
		name: "USDT",
		symbol: "USDT",
		decimals: "6",
		address: "0xFd086bC7CD5C481DCC9C85ebE478A1C0b69FCbb9",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "",
		lpResourceId: "",
		chainId: "42161",
		logoURI: "",
		mappedOnBridge: false,
		native: false,
		isLpToken: false,
		hasLpToken: false,
		stableAsset: true,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "USDC",
		symbol: "USDC",
		decimals: "6",
		address: "0xFF970A61A04b1cA14834A43f5dE4533eBDDB5CC8",
		lpSymbol: "RUSDC",
		lpAddress: "0x7bc1b2620fbc68c7435c07c812b2cd50b1d98ab9",
		stakingRewards: "",
		resourceId: "0x00000000000000000000002791bca1f2de4661ed88a30c99a7a9449aa8417400",
		lpResourceId: "0x0000000000000000000000510d0372f55a1687ce73e9826d33d7f3e43d849200",
		chainId: "42161",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: true,
		mining: false,
		activeMining: false,
		enableLiquidityMining: true
	},
	{
		name: "RUSDC",
		symbol: "RUSDC",
		decimals: "6",
		address: "0x7bc1b2620fbc68c7435c07c812b2cd50b1d98ab9",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000510d0372f55a1687ce73e9826d33d7f3e43d849200",
		lpResourceId: "",
		chainId: "42161",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: true,
		hasLpToken: false,
		stableAsset: true,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "DAI",
		symbol: "DAI",
		decimals: "18",
		address: "0xDA10009cBd5D07dd0CeCc66161FC93D7c9000da1",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "",
		lpResourceId: "",
		chainId: "42161",
		logoURI: "",
		mappedOnBridge: false,
		native: false,
		isLpToken: false,
		hasLpToken: false,
		stableAsset: true,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "ROUTE",
		symbol: "ROUTE",
		decimals: "18",
		address: "0x11BbF12363dC8375b78D2719395d505f52a02F68",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x000000000000000000000016eccfdbb4ee1a85a33f3a9b21175cd7ae753db400",
		lpResourceId: "",
		chainId: "42161",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "DFYN",
		symbol: "DFYN",
		decimals: "18",
		address: "0x13538f1450Ca2E1882Df650F87Eb996fF4Ffec34",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000c168e40227e4ebd8c1cae80f7a55a4f0e6d66c9700",
		lpResourceId: "",
		chainId: "42161",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "ETH",
		symbol: "ETH",
		decimals: "18",
		address: "0x82aF49447D8a07e3bd95BD0d56f35241523fBab1",
		lpSymbol: "RETH",
		lpAddress: "0x510d0372f55a1687ce73e9826d33d7f3e43d8492",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000c02aaa39b223fe8d0a0e5c4f27ead9083c756cc200",
		lpResourceId: "0x000000000000000000000064ffcaecb9a94569f2e082b260d2794168b96ac700",
		chainId: "42161",
		logoURI: "",
		mappedOnBridge: true,
		native: true,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: true
	},
	{
		name: "WETH",
		symbol: "WETH",
		decimals: "18",
		address: "0x82aF49447D8a07e3bd95BD0d56f35241523fBab1",
		lpSymbol: "RETH",
		lpAddress: "0x510d0372f55a1687ce73e9826d33d7f3e43d8492",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000c02aaa39b223fe8d0a0e5c4f27ead9083c756cc200",
		lpResourceId: "0x000000000000000000000064ffcaecb9a94569f2e082b260d2794168b96ac700",
		chainId: "42161",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "WMATIC",
		symbol: "WMATIC",
		decimals: "18",
		address: "0xfCAdfbAab63dd8ed485113eAf4dCc51CE68B1D3B",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x00000000000000000000004c28f48448720e9000907bc2611f73022fdce1fa00",
		lpResourceId: "",
		chainId: "42161",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "WAVAX",
		symbol: "WAVAX",
		decimals: "18",
		address: "0x71cB836342A9072d39fA100e2d44a982Aff4B54d",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000b31f66aa3c1e785363f0875a1b74e27b85fd66c700",
		lpResourceId: "",
		chainId: "42161",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "WFTM",
		symbol: "WFTM",
		decimals: "18",
		address: "0xCE2A671E7DE34a7B69233A14Da46A1Be7A91F00e",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x000000000000000000000021be370d5312f44cb42ce377bc9b8a0cef1a4c8300",
		lpResourceId: "",
		chainId: "42161",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "RETH",
		symbol: "RETH",
		decimals: "18",
		address: "0x510d0372f55a1687ce73e9826d33d7f3e43d8492",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x000000000000000000000064ffcaecb9a94569f2e082b260d2794168b96ac700",
		lpResourceId: "",
		chainId: "42161",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: true,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	}
];

var optimismAsset = [
	{
		name: "USDT",
		symbol: "USDT",
		decimals: "6",
		address: "0x94b008aA00579c1307B0EF2c499aD98a8ce58e58",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "",
		lpResourceId: "",
		chainId: "10",
		logoURI: "",
		mappedOnBridge: false,
		native: false,
		isLpToken: false,
		hasLpToken: false,
		stableAsset: true,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "USDC",
		symbol: "USDC",
		decimals: "6",
		address: "0x7F5c764cBc14f9669B88837ca1490cCa17c31607",
		lpSymbol: "RUSDC",
		lpAddress: "0xc3394bb9b118b3c9b5b8da297e20c08878edbed4",
		stakingRewards: "",
		resourceId: "0x00000000000000000000002791bca1f2de4661ed88a30c99a7a9449aa8417400",
		lpResourceId: "0x0000000000000000000000510d0372f55a1687ce73e9826d33d7f3e43d849200",
		chainId: "10",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: true,
		mining: false,
		activeMining: false,
		enableLiquidityMining: true
	},
	{
		name: "RUSDC",
		symbol: "RUSDC",
		decimals: "6",
		address: "0xc3394bb9b118b3c9b5b8da297e20c08878edbed4",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000510d0372f55a1687ce73e9826d33d7f3e43d849200",
		lpResourceId: "",
		chainId: "10",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: true,
		hasLpToken: false,
		stableAsset: true,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "DAI",
		symbol: "DAI",
		decimals: "18",
		address: "0xDA10009cBd5D07dd0CeCc66161FC93D7c9000da1",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "",
		lpResourceId: "",
		chainId: "10",
		logoURI: "",
		mappedOnBridge: false,
		native: false,
		isLpToken: false,
		hasLpToken: false,
		stableAsset: true,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "ROUTE",
		symbol: "ROUTE",
		decimals: "18",
		address: "0x8413041a7702603d9d991F2C4ADd29e4e8A241F8",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x000000000000000000000016eccfdbb4ee1a85a33f3a9b21175cd7ae753db400",
		lpResourceId: "",
		chainId: "10",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "DFYN",
		symbol: "DFYN",
		decimals: "18",
		address: "0xf44Ff799eA2bBFeC96f9A50498209AAc3C2b3b8b",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000c168e40227e4ebd8c1cae80f7a55a4f0e6d66c9700",
		lpResourceId: "",
		chainId: "10",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "ETH",
		symbol: "ETH",
		decimals: "18",
		address: "0x4200000000000000000000000000000000000006",
		lpSymbol: "RETH",
		lpAddress: "0x64ffcaecb9a94569f2e082b260d2794168b96ac7",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000c02aaa39b223fe8d0a0e5c4f27ead9083c756cc200",
		lpResourceId: "0x000000000000000000000064ffcaecb9a94569f2e082b260d2794168b96ac700",
		chainId: "10",
		logoURI: "",
		mappedOnBridge: true,
		native: true,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: true
	},
	{
		name: "WETH",
		symbol: "WETH",
		decimals: "18",
		address: "0x4200000000000000000000000000000000000006",
		lpSymbol: "RETH",
		lpAddress: "0x64ffcaecb9a94569f2e082b260d2794168b96ac7",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000c02aaa39b223fe8d0a0e5c4f27ead9083c756cc200",
		lpResourceId: "0x000000000000000000000064ffcaecb9a94569f2e082b260d2794168b96ac700",
		chainId: "10",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: true,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "WMATIC",
		symbol: "WMATIC",
		decimals: "18",
		address: "0x05bd0695aF629185fd2DCb13a99A3f0E7eF73943",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x00000000000000000000004c28f48448720e9000907bc2611f73022fdce1fa00",
		lpResourceId: "",
		chainId: "10",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "WAVAX",
		symbol: "WAVAX",
		decimals: "18",
		address: "0x5cAB1b7dDBA3a59D307e6f4b55b7329e5ABFB549",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x0000000000000000000000b31f66aa3c1e785363f0875a1b74e27b85fd66c700",
		lpResourceId: "",
		chainId: "10",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "WFTM",
		symbol: "WFTM",
		decimals: "18",
		address: "0x0C31c3649BbdB866809a28F6f4146a0c4c043FE9",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x000000000000000000000021be370d5312f44cb42ce377bc9b8a0cef1a4c8300",
		lpResourceId: "",
		chainId: "10",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: false,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	},
	{
		name: "RETH",
		symbol: "RETH",
		decimals: "18",
		address: "0x64ffcaecb9a94569f2e082b260d2794168b96ac7",
		lpSymbol: "",
		lpAddress: "",
		stakingRewards: "",
		resourceId: "0x000000000000000000000064ffcaecb9a94569f2e082b260d2794168b96ac700",
		lpResourceId: "",
		chainId: "10",
		logoURI: "",
		mappedOnBridge: true,
		native: false,
		isLpToken: true,
		hasLpToken: false,
		stableAsset: false,
		mining: false,
		activeMining: false,
		enableLiquidityMining: false
	}
];

var ChainId;

(function (ChainId) {
  ChainId["MAINNET"] = "1";
  ChainId["ROPSTEN"] = "3";
  ChainId["RINKEBY"] = "4";
  ChainId["G\u00D6RLI"] = "5";
  ChainId["KOVAN"] = "42";
  ChainId["POLYGON"] = "137";
  ChainId["MUMBAI"] = "80001";
  ChainId["OKEX"] = "66";
  ChainId["ARBITRUM"] = "42161";
  ChainId["FANTOM"] = "250";
  ChainId["OPTIMISM"] = "10";
  ChainId["XDAI"] = "100";
  ChainId["BSC"] = "56";
  ChainId["HARMONY"] = "1666600000";
  ChainId["AVALANCHE"] = "43114";
  ChainId["FUJI"] = "43113";
})(ChainId || (ChainId = {}));

const ASSET_LIST = {
  [ChainId.KOVAN]: kovanAsset,
  [ChainId.MUMBAI]: maticAsset,
  [ChainId.BSC]: bscAsset,
  [ChainId.AVALANCHE]: avaxAsset,
  [ChainId.FUJI]: avaxFujiAsset,
  [ChainId.MAINNET]: ethereumAsset,
  [ChainId.POLYGON]: polgonAsset,
  [ChainId.FANTOM]: ftmAsset,
  [ChainId.ARBITRUM]: arbitrumAsset,
  [ChainId.OPTIMISM]: optimismAsset,
  [ChainId.ROPSTEN]: ropstenAsset,
  [ChainId.OKEX]: okexAsset
};

const ROUTER_STATS_HOST = "https://api.stats.routerprotocol.com";
const NETWORK_CONFIG_URL = "https://raw.githubusercontent.com/router-protocol/sdk-network-config/main/networkConfig/mainnet.json";

const isAddressNative = (address, chainId) => {
  return NETWORK_CONFIG[chainId].native === address;
};

const PATH_FINDER_API_URL = "https://api.pathfinder.routerprotocol.com/api"; // calling the pathfinder api using axios

const fetchPathfinderData = async params => {
  const endpoint = "quote";
  const pathUrl = `${PATH_FINDER_API_URL}/${endpoint}`;

  try {
    const res = await axios.get(pathUrl, {
      params
    });
    return res.data;
  } catch (e) {
    console.error(`Fetching data from pathfinder: ${e}`);
  }
};

const getSwapQuote = async ({
  expandedAmount,
  sourceNetworkId,
  destinationNetworkId,
  sourceTokenAddress,
  destinationTokenAddress,
  userAddress,
  feeTokenAddress,
  slippageTolerance,
  network_config,
  widgetId
}) => {
  let args;

  if (sourceNetworkId === destinationNetworkId) {
    args = {
      'fromTokenAddress': sourceTokenAddress,
      'toTokenAddress': destinationTokenAddress,
      'amount': expandedAmount,
      'fromTokenChainId': sourceNetworkId,
      'toTokenChainId': destinationNetworkId,
      'userAddress': userAddress,
      'feeTokenAddress': feeTokenAddress,
      'slippageTolerance': slippageTolerance,
      widgetId
    }; // pathUrl = `${PATH_FINDER_ENDPOINT}?fromTokenAddress=${sourceTokenAddress}&toTokenAddress=${destinationTokenAddress}&amount=${expandedAmount}&fromTokenChainId=${sourceNetworkId}&toTokenChainId=${destinationNetworkId}`
  } else {
    if (isAddressNative(sourceTokenAddress, sourceNetworkId) || isAddressNative(destinationTokenAddress, destinationNetworkId)) {
      // pathUrl = `${PATH_FINDER_ENDPOINT}?fromTokenAddress=${network_config[sourceNetworkId].wnative}&toTokenAddress=${destinationTokenAddress}&amount=${expandedAmount}&fromTokenChainId=${sourceNetworkId}&toTokenChainId=${destinationNetworkId}`
      args = {
        'fromTokenAddress': network_config[sourceNetworkId].wnative,
        'toTokenAddress': network_config[destinationNetworkId].wnative,
        'amount': expandedAmount,
        'fromTokenChainId': sourceNetworkId,
        'toTokenChainId': destinationNetworkId,
        'userAddress': userAddress,
        'feeTokenAddress': feeTokenAddress,
        'slippageTolerance': slippageTolerance,
        widgetId
      };
    } else if (isAddressNative(sourceTokenAddress, sourceNetworkId)) {
      // pathUrl = `${PATH_FINDER_ENDPOINT}?fromTokenAddress=${network_config[sourceNetworkId].wnative}&toTokenAddress=${destinationTokenAddress}&amount=${expandedAmount}&fromTokenChainId=${sourceNetworkId}&toTokenChainId=${destinationNetworkId}`
      args = {
        'fromTokenAddress': network_config[sourceNetworkId].wnative,
        'toTokenAddress': destinationTokenAddress,
        'amount': expandedAmount,
        'fromTokenChainId': sourceNetworkId,
        'toTokenChainId': destinationNetworkId,
        'userAddress': userAddress,
        'feeTokenAddress': feeTokenAddress,
        'slippageTolerance': slippageTolerance,
        widgetId
      };
    } else if (isAddressNative(destinationTokenAddress, destinationNetworkId)) {
      // pathUrl = `${PATH_FINDER_ENDPOINT}?fromTokenAddress=${sourceTokenAddress}&toTokenAddress=${network_config[destinationNetworkId].wnative}&amount=${expandedAmount}&fromTokenChainId=${sourceNetworkId}&toTokenChainId=${destinationNetworkId}`
      args = {
        'fromTokenAddress': sourceTokenAddress,
        'toTokenAddress': network_config[destinationNetworkId].wnative,
        'amount': expandedAmount,
        'fromTokenChainId': sourceNetworkId,
        'toTokenChainId': destinationNetworkId,
        'userAddress': userAddress,
        'feeTokenAddress': feeTokenAddress,
        'slippageTolerance': slippageTolerance,
        widgetId
      };
    } else {
      // pathUrl = `${PATH_FINDER_ENDPOINT}?fromTokenAddress=${sourceTokenAddress}&toTokenAddress=${destinationTokenAddress}&amount=${expandedAmount}&fromTokenChainId=${sourceNetworkId}&toTokenChainId=${destinationNetworkId}`
      args = {
        'fromTokenAddress': sourceTokenAddress,
        'toTokenAddress': destinationTokenAddress,
        'amount': expandedAmount,
        'fromTokenChainId': sourceNetworkId,
        'toTokenChainId': destinationNetworkId,
        'userAddress': userAddress,
        'feeTokenAddress': feeTokenAddress,
        'slippageTolerance': slippageTolerance,
        widgetId
      };
    }
  }

  try {
    if (args) {
      const pathfinder_response = await fetchPathfinderData(args);
      return pathfinder_response;
    }
  } catch (e) {
    return e;
  }
};

var erc20Token = [
	{
		constant: true,
		inputs: [
		],
		name: "name",
		outputs: [
			{
				name: "",
				type: "string"
			}
		],
		payable: false,
		stateMutability: "view",
		type: "function"
	},
	{
		constant: false,
		inputs: [
			{
				name: "_spender",
				type: "address"
			},
			{
				name: "_value",
				type: "uint256"
			}
		],
		name: "approve",
		outputs: [
			{
				name: "",
				type: "bool"
			}
		],
		payable: false,
		stateMutability: "nonpayable",
		type: "function"
	},
	{
		constant: true,
		inputs: [
		],
		name: "totalSupply",
		outputs: [
			{
				name: "",
				type: "uint256"
			}
		],
		payable: false,
		stateMutability: "view",
		type: "function"
	},
	{
		constant: false,
		inputs: [
			{
				name: "_from",
				type: "address"
			},
			{
				name: "_to",
				type: "address"
			},
			{
				name: "_value",
				type: "uint256"
			}
		],
		name: "transferFrom",
		outputs: [
			{
				name: "",
				type: "bool"
			}
		],
		payable: false,
		stateMutability: "nonpayable",
		type: "function"
	},
	{
		constant: true,
		inputs: [
		],
		name: "decimals",
		outputs: [
			{
				name: "",
				type: "uint8"
			}
		],
		payable: false,
		stateMutability: "view",
		type: "function"
	},
	{
		constant: true,
		inputs: [
			{
				name: "_owner",
				type: "address"
			}
		],
		name: "balanceOf",
		outputs: [
			{
				name: "balance",
				type: "uint256"
			}
		],
		payable: false,
		stateMutability: "view",
		type: "function"
	},
	{
		constant: true,
		inputs: [
		],
		name: "symbol",
		outputs: [
			{
				name: "",
				type: "string"
			}
		],
		payable: false,
		stateMutability: "view",
		type: "function"
	},
	{
		constant: false,
		inputs: [
			{
				name: "_to",
				type: "address"
			},
			{
				name: "_value",
				type: "uint256"
			}
		],
		name: "transfer",
		outputs: [
			{
				name: "",
				type: "bool"
			}
		],
		payable: false,
		stateMutability: "nonpayable",
		type: "function"
	},
	{
		constant: true,
		inputs: [
			{
				name: "_owner",
				type: "address"
			},
			{
				name: "_spender",
				type: "address"
			}
		],
		name: "allowance",
		outputs: [
			{
				name: "",
				type: "uint256"
			}
		],
		payable: false,
		stateMutability: "view",
		type: "function"
	},
	{
		payable: true,
		stateMutability: "payable",
		type: "fallback"
	},
	{
		anonymous: false,
		inputs: [
			{
				indexed: true,
				name: "owner",
				type: "address"
			},
			{
				indexed: true,
				name: "spender",
				type: "address"
			},
			{
				indexed: false,
				name: "value",
				type: "uint256"
			}
		],
		name: "Approval",
		type: "event"
	},
	{
		anonymous: false,
		inputs: [
			{
				indexed: true,
				name: "from",
				type: "address"
			},
			{
				indexed: true,
				name: "to",
				type: "address"
			},
			{
				indexed: false,
				name: "value",
				type: "uint256"
			}
		],
		name: "Transfer",
		type: "event"
	}
];

const getSourceTokenApproval = async ({
  tokenAddress,
  sourceNetworkId,
  destinationNetworkId,
  userAddress,
  provider,
  network_config // spenderAddress, // Cross chain reserve handler / same chain one-split

}) => {
  try {
    let spenderAddress;

    if (sourceNetworkId === destinationNetworkId) {
      spenderAddress = network_config[sourceNetworkId].oneSplit;
    } else {
      spenderAddress = network_config[sourceNetworkId].reserveHandler;
    }

    const erc20Instance = new ethers.Contract(tokenAddress, erc20Token, provider);
    const result = await erc20Instance.allowance(userAddress, spenderAddress);
    return result;
  } catch (error) {
    console.log("Error", error);
  }
};
const getFeeTokenApproval = async ({
  tokenAddress,
  networkId,
  userAddress,
  provider,
  network_config
}) => {
  try {
    const spenderAddress = network_config[networkId].reserveHandler;
    const erc20Instance = new ethers.Contract(tokenAddress, erc20Token, provider);
    const result = await erc20Instance.allowance(userAddress, spenderAddress);
    return result;
  } catch (error) {
    console.log("Error", error);
  }
};
const setSourceTokenApproval = async ({
  tokenAddress,
  userAddress,
  amount,
  sourceNetworkId,
  destinationNetworkId,
  network_config
}) => {
  let spenderAddress;

  if (sourceNetworkId === destinationNetworkId) {
    spenderAddress = network_config[sourceNetworkId].oneSplit;
  } else {
    spenderAddress = network_config[sourceNetworkId].reserveHandler;
  }

  try {
    const iface = new Interface(erc20Token);
    const rawData = iface.encodeFunctionData("approve", [spenderAddress, amount]);
    return {
      to: tokenAddress,
      from: userAddress,
      data: rawData
    };
  } catch (error) {
    console.log("Error", error);
    return undefined;
  }
};
const setFeeTokenApproval = async ({
  amount,
  tokenAddress,
  userAddress,
  sourceNetworkId,
  network_config
}) => {
  const spenderAddress = network_config[sourceNetworkId].reserveHandler;

  try {
    const iface = new Interface(erc20Token);
    const rawData = iface.encodeFunctionData("approve", [spenderAddress, amount]);
    return {
      to: tokenAddress,
      from: userAddress,
      data: rawData
    };
  } catch (error) {
    console.log("Error", error);
    return undefined;
  }
};

var feeHandlerabi = [
	{
		anonymous: false,
		inputs: [
			{
				indexed: false,
				internalType: "uint8",
				name: "destinationChainID",
				type: "uint8"
			},
			{
				indexed: false,
				internalType: "address",
				name: "feeTokenAddress",
				type: "address"
			},
			{
				indexed: false,
				internalType: "uint256",
				name: "transferFee",
				type: "uint256"
			},
			{
				indexed: false,
				internalType: "uint256",
				name: "exchangeFee",
				type: "uint256"
			},
			{
				indexed: false,
				internalType: "bool",
				name: "accepted",
				type: "bool"
			}
		],
		name: "FeeUpdated",
		type: "event"
	},
	{
		anonymous: false,
		inputs: [
			{
				indexed: true,
				internalType: "bytes32",
				name: "role",
				type: "bytes32"
			},
			{
				indexed: true,
				internalType: "bytes32",
				name: "previousAdminRole",
				type: "bytes32"
			},
			{
				indexed: true,
				internalType: "bytes32",
				name: "newAdminRole",
				type: "bytes32"
			}
		],
		name: "RoleAdminChanged",
		type: "event"
	},
	{
		anonymous: false,
		inputs: [
			{
				indexed: true,
				internalType: "bytes32",
				name: "role",
				type: "bytes32"
			},
			{
				indexed: true,
				internalType: "address",
				name: "account",
				type: "address"
			},
			{
				indexed: true,
				internalType: "address",
				name: "sender",
				type: "address"
			}
		],
		name: "RoleGranted",
		type: "event"
	},
	{
		anonymous: false,
		inputs: [
			{
				indexed: true,
				internalType: "bytes32",
				name: "role",
				type: "bytes32"
			},
			{
				indexed: true,
				internalType: "address",
				name: "account",
				type: "address"
			},
			{
				indexed: true,
				internalType: "address",
				name: "sender",
				type: "address"
			}
		],
		name: "RoleRevoked",
		type: "event"
	},
	{
		inputs: [
		],
		name: "DEFAULT_ADMIN_ROLE",
		outputs: [
			{
				internalType: "bytes32",
				name: "",
				type: "bytes32"
			}
		],
		stateMutability: "view",
		type: "function"
	},
	{
		inputs: [
		],
		name: "fetchHandler",
		outputs: [
			{
				internalType: "address",
				name: "",
				type: "address"
			}
		],
		stateMutability: "view",
		type: "function"
	},
	{
		inputs: [
			{
				internalType: "uint8",
				name: "destChainId",
				type: "uint8"
			}
		],
		name: "getChainFeeTokens",
		outputs: [
			{
				internalType: "address[]",
				name: "",
				type: "address[]"
			}
		],
		stateMutability: "view",
		type: "function"
	},
	{
		inputs: [
			{
				internalType: "uint8",
				name: "destinationChainID",
				type: "uint8"
			},
			{
				internalType: "address",
				name: "feeTokenAddress",
				type: "address"
			}
		],
		name: "getFee",
		outputs: [
			{
				internalType: "uint256",
				name: "",
				type: "uint256"
			},
			{
				internalType: "uint256",
				name: "",
				type: "uint256"
			}
		],
		stateMutability: "view",
		type: "function"
	},
	{
		inputs: [
			{
				internalType: "uint8",
				name: "destChainId",
				type: "uint8"
			},
			{
				internalType: "address",
				name: "feeToken",
				type: "address"
			}
		],
		name: "getFeeSafe",
		outputs: [
			{
				internalType: "uint256",
				name: "",
				type: "uint256"
			},
			{
				internalType: "uint256",
				name: "",
				type: "uint256"
			},
			{
				internalType: "bool",
				name: "",
				type: "bool"
			}
		],
		stateMutability: "view",
		type: "function"
	},
	{
		inputs: [
			{
				internalType: "bytes32",
				name: "role",
				type: "bytes32"
			}
		],
		name: "getRoleAdmin",
		outputs: [
			{
				internalType: "bytes32",
				name: "",
				type: "bytes32"
			}
		],
		stateMutability: "view",
		type: "function"
	},
	{
		inputs: [
			{
				internalType: "bytes32",
				name: "role",
				type: "bytes32"
			},
			{
				internalType: "address",
				name: "account",
				type: "address"
			}
		],
		name: "grantRole",
		outputs: [
		],
		stateMutability: "nonpayable",
		type: "function"
	},
	{
		inputs: [
			{
				internalType: "bytes32",
				name: "role",
				type: "bytes32"
			},
			{
				internalType: "address",
				name: "account",
				type: "address"
			}
		],
		name: "hasRole",
		outputs: [
			{
				internalType: "bool",
				name: "",
				type: "bool"
			}
		],
		stateMutability: "view",
		type: "function"
	},
	{
		inputs: [
			{
				internalType: "address",
				name: "handlerAddress",
				type: "address"
			}
		],
		name: "initialize",
		outputs: [
		],
		stateMutability: "nonpayable",
		type: "function"
	},
	{
		inputs: [
			{
				internalType: "bytes32",
				name: "role",
				type: "bytes32"
			},
			{
				internalType: "address",
				name: "account",
				type: "address"
			}
		],
		name: "renounceRole",
		outputs: [
		],
		stateMutability: "nonpayable",
		type: "function"
	},
	{
		inputs: [
			{
				internalType: "bytes32",
				name: "role",
				type: "bytes32"
			},
			{
				internalType: "address",
				name: "account",
				type: "address"
			}
		],
		name: "revokeRole",
		outputs: [
		],
		stateMutability: "nonpayable",
		type: "function"
	},
	{
		inputs: [
			{
				internalType: "uint8",
				name: "destinationChainID",
				type: "uint8"
			},
			{
				internalType: "address",
				name: "feeTokenAddress",
				type: "address"
			},
			{
				internalType: "uint256",
				name: "transferFee",
				type: "uint256"
			},
			{
				internalType: "uint256",
				name: "exchangeFee",
				type: "uint256"
			},
			{
				internalType: "bool",
				name: "accepted",
				type: "bool"
			}
		],
		name: "setFee",
		outputs: [
		],
		stateMutability: "nonpayable",
		type: "function"
	},
	{
		inputs: [
			{
				internalType: "address",
				name: "_handler",
				type: "address"
			}
		],
		name: "setHandler",
		outputs: [
		],
		stateMutability: "nonpayable",
		type: "function"
	},
	{
		inputs: [
			{
				internalType: "bytes4",
				name: "interfaceId",
				type: "bytes4"
			}
		],
		name: "supportsInterface",
		outputs: [
			{
				internalType: "bool",
				name: "",
				type: "bool"
			}
		],
		stateMutability: "view",
		type: "function"
	},
	{
		inputs: [
			{
				internalType: "address",
				name: "tokenAddress",
				type: "address"
			},
			{
				internalType: "address",
				name: "recipient",
				type: "address"
			},
			{
				internalType: "uint256",
				name: "amount",
				type: "uint256"
			}
		],
		name: "withdrawFee",
		outputs: [
		],
		stateMutability: "nonpayable",
		type: "function"
	}
];

const getFeeFromFeeHandlerMulticaller = ({
  feeHandlerAddress,
  dstChainID,
  feeTokenAddress
}) => {
  const feeHandlerInstance = new Contract(feeHandlerAddress, feeHandlerabi);
  return feeHandlerInstance.getFeeSafe(dstChainID, feeTokenAddress);
};

const getFeeTokens = async ({
  feeHandlerAddress,
  provider,
  dstChainID
}) => {
  const feeHandlerInstance = new ethers.Contract(feeHandlerAddress, feeHandlerabi, provider);

  try {
    const tokens = await feeHandlerInstance.getChainFeeTokens(dstChainID);
    return tokens;
  } catch (e) {
    console.log("Fee handler error - ", e);
  }
}; //return (fees._transferFee, fees._exchangeFee, fees.accepted);


const getFeeAmount = async ({
  sourceProvider,
  sourceNetworkId,
  destNetworkId,
  network_config
}) => {
  const ethcallProvider = new Provider(sourceProvider, Number(sourceNetworkId));
  const sourceFeeHandlerAddress = network_config[sourceNetworkId].feeHandler;
  const destChainId = network_config[destNetworkId].id;
  const feeTokenList = await getFeeTokens({
    feeHandlerAddress: sourceFeeHandlerAddress,
    provider: sourceProvider,
    dstChainID: destChainId
  });
  const contractCalls = feeTokenList.map(feeTokenAddress => getFeeFromFeeHandlerMulticaller({
    feeHandlerAddress: sourceFeeHandlerAddress,
    dstChainID: destChainId,
    feeTokenAddress: feeTokenAddress
  }));
  const feeArray = await ethcallProvider.all(contractCalls);
  const newFeeTokenList = [];
  feeTokenList.forEach((feeTokenAddress, index) => {
    if (feeArray[index][2]) {
      const feeObject = {
        address: feeTokenAddress,
        transferFee: feeArray[index][0],
        exchangeFee: feeArray[index][1]
      };
      newFeeTokenList.push(feeObject);
    }
  });
  return newFeeTokenList;
};

const getTxStatus = async (txHash, networkId) => {
  const params = {
    txHash,
    networkId
  };
  const pathUrl = `${ROUTER_STATS_HOST}/api/status`;

  try {
    const res = await axios.get(pathUrl, {
      params
    });
    return res.data;
  } catch (e) {
    console.error(`Fetching Transaction Status: ${e}`);
    return undefined;
  }
};

class RouterProtocol {
  constructor(apiKey, sourceNetworkId, provider) {
    this.apiKey = apiKey;
    this.provider = provider;
    this.sourceNetworkId = sourceNetworkId;
  }

  async initialize() {
    try {
      const networks = await axios.get(NETWORK_CONFIG_URL).then(res => {
        return res.data;
      });
      let network_config = {};
      networks.forEach(network => {
        network_config = { ...network_config,
          [network.networkId]: network
        };
        return;
      });
      this.networkConfig = network_config;
    } catch (err) {
      throw err;
    }
  } // to-be deprecated


  async initailize() {
    try {
      const networks = await axios.get(NETWORK_CONFIG_URL).then(res => {
        return res.data;
      });
      let network_config = {};
      networks.forEach(network => {
        network_config = { ...network_config,
          [network.networkId]: network
        };
        return;
      });
      this.networkConfig = network_config;
    } catch (err) {
      throw err;
    }
  }
  /**
   * gets quote
   */


  async getQuote(amount, destinationNetworkId, sourceTokenAddress, destinationTokenAddress, userAddress, feeTokenAddress, slippageTolerance) {
    try {
      if (!this.networkConfig) throw "Not initialized!";
      const res = await getSwapQuote({
        expandedAmount: amount,
        sourceNetworkId: this.sourceNetworkId,
        destinationNetworkId,
        sourceTokenAddress,
        destinationTokenAddress,
        userAddress,
        feeTokenAddress,
        slippageTolerance,
        network_config: this.networkConfig,
        widgetId: this.apiKey
      });
      return res;
    } catch (err) {
      throw err;
    }
  }
  /**
   * swaps
   */


  async swap(quote, signer, txGasLimit) {
    try {
      if (!this.networkConfig) throw "Not initialized!";
      if (!quote) throw "No Quote"; // setting the gas price and limit

      const gasPrice = await this.provider.getGasPrice();
      const gasLimit = ethers.utils.hexlify(Number(txGasLimit ? txGasLimit : 1000000));
      const transaction = await signer.sendTransaction({ ...quote.txn.execution,
        gasPrice,
        gasLimit
      });
      await transaction.wait();
      return transaction;
    } catch (err) {
      throw err;
    }
  }

  async getSourceTokenAllowance(tokenAddress, destinationNetworkId, userAddress) {
    try {
      if (!this.networkConfig) throw "Not initialized!";
      return await getSourceTokenApproval({
        tokenAddress,
        sourceNetworkId: this.sourceNetworkId,
        provider: this.provider,
        destinationNetworkId,
        userAddress,
        network_config: this.networkConfig
      });
    } catch (err) {
      throw err;
    }
  }

  async getFeeTokenAllowance(tokenAddress, destinationNetworkId, userAddress) {
    try {
      if (!this.networkConfig) throw "Not initialized!";
      return await getFeeTokenApproval({
        tokenAddress,
        networkId: destinationNetworkId,
        provider: this.provider,
        userAddress,
        network_config: this.networkConfig
      });
    } catch (err) {
      throw err;
    }
  }

  async approveSourceToken(tokenAddress, userAddress, amount, destinationNetworkId, signer, txGasLimit) {
    try {
      if (!this.networkConfig) throw "Not initialized!";
      const transactionReq = await setSourceTokenApproval({
        tokenAddress,
        amount,
        destinationNetworkId,
        userAddress,
        sourceNetworkId: this.sourceNetworkId,
        network_config: this.networkConfig
      }); // setting the gas price and limit

      const gasPrice = await this.provider.getGasPrice();
      const gasLimit = ethers.utils.hexlify(Number(txGasLimit ? txGasLimit : 1000000));
      const transaction = await signer.sendTransaction({ ...transactionReq,
        gasPrice,
        gasLimit
      });
      await transaction.wait();
      return transaction;
    } catch (err) {
      throw err;
    }
  }

  async approveFeeToken(tokenAddress, userAddress, amount, signer, txGasLimit) {
    try {
      if (!this.networkConfig) throw "Not initialized!";
      const transactionReq = await setFeeTokenApproval({
        tokenAddress,
        amount,
        userAddress,
        sourceNetworkId: this.sourceNetworkId,
        network_config: this.networkConfig
      }); // setting the gas price and limit

      const gasPrice = await this.provider.getGasPrice();
      const gasLimit = ethers.utils.hexlify(Number(txGasLimit ? txGasLimit : 1000000));
      const transaction = await signer.sendTransaction({ ...transactionReq,
        gasPrice,
        gasLimit
      });
      await transaction.wait();
      return transaction;
    } catch (err) {
      throw err;
    }
  }

  async getBridgeFee(destinationNetworkId) {
    try {
      if (!this.networkConfig) throw "Not initialized!";
      const fee = await getFeeAmount({
        sourceNetworkId: this.sourceNetworkId,
        destNetworkId: destinationNetworkId,
        sourceProvider: this.provider,
        network_config: this.networkConfig
      });
      return fee;
    } catch (err) {
      throw err;
    }
  }

  async getTransactionStatus(txHash) {
    try {
      if (!this.networkConfig) throw "Not initialized!";
      const res = await getTxStatus(txHash, this.sourceNetworkId);
      return res;
    } catch (err) {
      throw err;
    }
  }

} // whole test
// feeTokenName
// transaction status in midway
// transaction status endpoint

export { ASSET_LIST, ChainId, NETWORK_CONFIG, NETWORK_CONFIG_URL, ROUTER_STATS_HOST, RouterProtocol };
